src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6338,120 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
             {
                 return this.ParseTupleType();
             }
+            else if (IsFunctionPointerStart())
+            {
+                return ParseFunctionPointerTypeSyntax();
+            }
             else
             {
                 var name = this.CreateMissingIdentifierName();
                 return this.AddError(name, ErrorCode.ERR_TypeExpected);
             }
         }
 
+        private FunctionPointerTypeSyntax ParseFunctionPointerTypeSyntax()
+        {
+            Debug.Assert(IsFunctionPointerStart());
+            var @delegate = EatToken(SyntaxKind.DelegateKeyword);
+            var asterisk = EatToken(SyntaxKind.AsteriskToken);
+            var callingConvention = CurrentToken.Kind != SyntaxKind.LessThanToken
+                ? CurrentToken.ContextualKind switch
+                {
+                    SyntaxKind.CdeclKeyword => EatContextualToken(SyntaxKind.CdeclKeyword),
+                    SyntaxKind.ManagedKeyword => EatContextualToken(SyntaxKind.ManagedKeyword),
+                    SyntaxKind.UnmanagedKeyword => EatContextualToken(SyntaxKind.UnmanagedKeyword),
+                    SyntaxKind.ThiscallKeyword => EatContextualToken(SyntaxKind.ThiscallKeyword),
+                    SyntaxKind.StdcallKeyword => EatContextualToken(SyntaxKind.StdcallKeyword),
+                    _ => parseInvalidCallingConvention(),
+                }
+                : null;
```

Comment:
`var callingConvention = kind != lessthan ? EatCallingConvention() : null;`

Then define `EatCallingConvention` below (as local or other instance method).   Interspersing a large pattern like this just feels difficult to read and grasp.

Also, all your branches appear the same.  Basically, `ConvertToKeyword(this.EatToken())`.  So it feels like a normal switch where those 5 cases call the same thing is clearer and less error prone.

Reply:
you could also just have all the code be in its own local function (since you already do that for parseInvalidCallingConv).  The mixing/matching of styles/logic here isn't great IMO.